rr_evgp/src/costmap_plugins/track_closing_layer.cpp

```diff
@@ -0,0 +1,248 @@
+    assertions::getParam(private_nh, "reg_bounding_box_size", reg_bounding_box_size_, { assertions::greater<int>(0) });
+    assertions::getParam(private_nh, "extrapolate_distance", extrapolate_distance_, { assertions::greater<double>(0) });
+    assertions::getParam(private_nh, "max_angle_between_lines", max_angle_between_lines_,
+                         { assertions::greater<double>(0) });
+    assertions::getParam(private_nh, "max_dist_between_lines", max_dist_between_lines_,
+                         { assertions::greater<double>(0) });
+}
+
+void TrackClosingLayer::updateBounds(double robot_x, double robot_y, double robot_yaw, double* min_x, double* min_y,
+                                     double* max_x, double* max_y) {
+    if (!enabled_) {
+        return;
+    }
+
+    const costmap_2d::Costmap2D* costmap = layered_costmap_->getCostmap();
+    *min_x = std::min(*min_x, costmap->getOriginX());
+    *min_y = std::min(*min_y, costmap->getOriginY());
+    *max_x = std::max(*max_x, costmap->getOriginX() + costmap->getSizeInMetersX());
+    *max_y = std::max(*max_y, costmap->getOriginY() + costmap->getSizeInMetersY());
+}
+
+template <int N>
+double norm(const cv::Vec<double, N>& x) {
+    return std::sqrt(x.dot(x));
+}
+
+cv::Point2d intersect_point_line(const cv::Point2d& p, const cv::Point2d& l1, const cv::Point2d& l2, bool bound) {
+    cv::Vec2d n = l2 - l1;
+    n /= norm(n);
+    double t = (p - l1).dot(n);
+    cv::Vec2d x = cv::Vec2d(l1) + n * t;
+
+    if (bound && (norm(x - cv::Vec2d(l1)) + norm(x - cv::Vec2d(l2)) > norm(l2 - l1))) {
+        x = (norm(p - l1) < norm(p - l2)) ? l1 : l2;
+    }
+
+    return x;
+}
+
+bool TrackClosingLayer::IsConnectable(const WallEndpointContext& current, const WallEndpointContext& prev) {
+    const auto& [endpntA, pntA, vxA, vyA] = current;
+    const auto& [endpntB, pntB, vxB, vyB] = prev;
+
+    double angle_between_lines = std::acos(vxA * vxB + vyA * vyB) * 180.0 / M_PI;
+    angle_between_lines = std::min(angle_between_lines, 180 - angle_between_lines);
+
+    cv::Point closest_endB_to_lineA = intersect_point_line(endpntB, pntA, pntA + cv::Point(vxA * 500, vyA * 500), true);
+    cv::Point closest_endA_to_lineB = intersect_point_line(endpntA, pntB, pntB + cv::Point(vxB * 500, vyB * 500), true);
+    double dist_a = norm(closest_endA_to_lineB - endpntA);
+    double dist_b = norm(closest_endB_to_lineA - endpntB);
+
+    return angle_between_lines <= max_angle_between_lines_ && dist_a < max_dist_between_lines_ &&
+           dist_b < max_dist_between_lines_;
+}
+
+void TrackClosingLayer::updateCosts(costmap_2d::Costmap2D& master_grid, int min_cell_x, int min_cell_y, int max_cell_x,
+                                    int max_cell_y) {
+    if (!enabled_) {
+        return;
+    }
+
+    // populate OpenCV Mat as grid
+    cv::Mat mat_grid(master_grid.getSizeInCellsX(), master_grid.getSizeInCellsY(), CV_8UC1);
+    uint8_t* charMap = master_grid.getCharMap();
```

Comment:
It's possible that the obstacle layer may edit masterGrid's underlying charMap asynchronously (something like this happens in ROS Kinetic). If so, we will need to add synchronization. So far I haven't seen the weird effects that I ran up against this summer, so that's good

Reply:
Going to leave this comment in the review, but won't fix it right now because it is not apparent what (if anything) needs to change